src/rgw/rgw_op.cc

```diff
@@ -7654,7 +7654,7 @@ int RGWHandler::do_read_permissions(RGWOp *op, bool only_bucket)
 		      << " ret=" << ret << dendl;
     if (ret == -ENODATA)
       ret = -EACCES;
-    if (s->auth.identity->is_anonymous() && ret == -EACCES)
+    if (s->auth.identity->is_anonymous() && (ret == -EACCES || ret == -ENOENT))
```

Comment:
where was the ENOENT coming from? looking at the tempest test case from https://opendev.org/openstack/tempest/src/branch/master/tempest/scenario/test_object_storage_basic_ops.py#L65, the object under test should exist there

doesn't this change mean that an anonymous user will never see a 404 error? surely that's not right

Reply:
It seemed to be coming when keystone auth was used. I'll look a bit closer into it and find out exactly.

I guess it depends on what a anonymous user would expect. I guess a 404 is ok if a container is readable to them. Otherwise it should be a 401. I guess in the former case, they wont get that anymore.

I'll delve a little deeper.